ext/opcache/zend_shared_alloc: add assertions on "locked" flag

MaxKellermann opened the pull request:
Let the PHP process crash if a bug causes incorrect locking calls.

cmb69 commented on ext/opcache/zend_shared_alloc.c:
```diff
@@ -336,6 +336,8 @@ static size_t zend_shared_alloc_get_largest_free_block(void)
 
 void *zend_shared_alloc(size_t size)
 {
+	ZEND_ASSERT(ZCG(locked));
```
I think this assertion makes sense, although a few lines down that condition is checked, and fatal error is triggered if there is no log. Maybe we should remove the preprocessor guard ('#if 1') for that, though.

MaxKellermann replied:
Oh yes, this `#if 1` is confusing - though I'd prefer to remove this section and use only `ZEND_ASSERT`, because this is not a fatal runtime error, but a bug, and that's what assertions are for.

cmb69 replied:
> because this is not a fatal runtime error, but a bug, and that's what assertions are for.

Ah, indeed!

cmb69 reviewed (commented):
Thank you! Looks generally good to me; only the first added assertion might need some thought.

MaxKellermann commented:
@cmb69, I moved the `zend_shared_alloc()` change to a separate commit and removed the runtime check, as mentioned in https://github.com/php/php-src/pull/10205#discussion_r1060019377
As a side effect, this check is eliminated in release builds (`-DNDEBUG`), but that's a good thing IMHO.
Agree?

cmb69 reviewed (approved):
Thank you!

MaxKellermann commented:
The MACOS_DEBUG_NTS failure looks spurious, not related to this PR?

cmb69 commented:
Yep, that test failure is definitely unrelated.
